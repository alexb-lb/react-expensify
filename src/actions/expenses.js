import database from '../firebase/firebase';

// component calls action generator
// action generator returns function
// component dispatches function (?)
// function runs (has the ability to dispath other actions and do whatever it wants)

// Add expense
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  // return function to access dispatch method as argument from redux store
  return (dispatch) => {
    // destructuring properties from expenseData
    // const for every name in obj (const description = ''; const note = '')
    const {description = '', note = '', amount = 0, createdAt = 0} = expenseData;

    const expense = {description, note, amount, createdAt};
    // push into Firebase and get back generated ID of expense, return promise
    return database.ref('expenses').push(expense).then((ref) => {
      // dispatch expense into redux store by passing an expense into addExpense func
      dispatch(addExpense({id: ref.key, ...expense}))
    })
  };
};

// Remove expense
export const removeExpense = ({id} = {}) => {
  return ({type: 'REMOVE_EXPENSE', id})
};

// Edit expense
export const editExpense = (id, updates = {}) => {
  return ({type: 'EDIT_EXPENSE', id, updates});
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  // return function to access dispatch method as argument from redux store
  return (dispatch) => {
    // return function to access promise chaining further
    return database.ref(`expenses`).once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push(({
          id: childSnapshot.key,
          ...childSnapshot.val()
        }))
      });
      // dispatch expense into redux store by passing an expense into addExpense func
      dispatch(setExpenses(expenses));
    });
  };
};


