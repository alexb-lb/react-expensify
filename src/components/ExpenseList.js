import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        props.expenses.map((expense, i, arr) => {
          return <ExpenseListItem key={expense.id} {...expense}/>
        })
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
};

// сначала передаем функ, которая достает данные из Redux Store
// потом указываем, куда эти данные нужно вывести
export default connect(mapStateToProps)(ExpenseList);