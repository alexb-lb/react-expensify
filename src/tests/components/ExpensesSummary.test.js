import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpenseSummary correctly', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={0} expensesTotal={[]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary correctly', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={26} expensesTotal={6123123231}/>);
  expect(wrapper).toMatchSnapshot();
});