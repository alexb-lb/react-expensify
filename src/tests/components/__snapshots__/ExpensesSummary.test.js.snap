// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
  </h1>
</div>
`;

exports[`should render ExpenseSummary correctly 2`] = `
<div>
  <h1>
    Viewing 
    26
     
    expenses
     totalling 
    $61,231,232.31
  </h1>
</div>
`;
